### PR TITLE
Esp config

### PR DESCRIPTION
**Pull request by Gretel5X**
Added complete new configuration capabilities. You will have to run `idf.py erase_flash` to before flashing the device. 
Important to test: 
- provision device via bluetooth and see that wifi credentials appear in the config even after reboot
- no error messages in the dev console for frontend, play with the settings, try enter weird stuff

The PR is large, pay close attention to formatting ;)
Until the pipeline can run with espressif framework version 4.3, the CI will fail

### CODE REVIEW

File: main/Kconfig.projbuild [hunk 1]
```diff
@@ -100,22 +91,18 @@ menu "User Configuration"
...
         string "URL of Emoncms post API endpoint"
-        depends on EMONCMS
         default "/emoncms/input/post"
 
     config EMONCMS_APIKEY
         string "API key for Emoncms access"
-        depends on EMONCMS
         default ""
 
     config EMONCMS_NODE_SERIAL
```
Comment (martinjaeger) [inline main/Kconfig.projbuild:108]: `depends on EMONCMS` has to be removed here as well, otherwise I get a compile error.

**Review by martinjaeger — COMMENTED** [review]
I'm getting this after the first start:
image

It's also not possible to select the actual device anymore. At the top I can see the list with "self" and "BMS 15S80 IC", but I can't select the BMS.

Any hints what I might be doing wrong?

File: main/config_nodes.h [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+extern "C"
+{
+#endif
+
+#include <stdint.h>
+#include "esp_err.h"
+#include "ts_client.h"
+
+
+#define CONFIG_LIBRE_SOLAR_TYPE_ID 9
```
Comment (martinjaeger) [inline main/config_nodes.h]: Namespace `CONFIG_` is reserved for auto-generated defines from Kconfig. Please remove the prefix.

File: main/config_nodes.h [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+#define ID_AUTH     0xEA
+#define ID_PUB      0xF0        // publication setup
+#define ID_SUB      0xF1        // subscription setup
+#define ID_LOG      0x100       // access log data
+
+#define STRING_LEN 128          // size of allocated strings in config
+#define CONFIG_NODE_NAME        "conf"
+#define CONFIG_NODE_GENERAL     "general"
+#define CONFIG_NODE_EMONCMS     "emoncms"
+#define CONFIG_NODE_MQTT        "mqtt"
```
Comment (martinjaeger) [inline main/config_nodes.h]: Prefix should be changed also here. Maybe replace with `CONF_`? Or why not keep previous naming "data node" as in charge controller?

**Comment by Gretel5X** [comment]
> I'm getting this after the first start:
> image
> 
> It's also not possible to select the actual device anymore. At the top I can see the list with "self" and "BMS 15S80 IC", but I can't select the BMS.
> 
> Any hints what I might be doing wrong?

Run `npm run build` to build the fresh frontend files prior to flashing the esp

File: main/config_nodes.h (hunk 2 above)
Comment (Gretel5X) [reply main/config_nodes.h:20]: I would move it to the defaults?

File: main/config_nodes.h (hunk 3 above)
Comment (Gretel5X) [reply main/config_nodes.h]: Yes good point, it turned out as a normal thingset setup so its fine :)

**Comment by martinjaeger** [comment]
Finally found time to play around with it. Looks really nice!

With the BMS that is currently attached I can't access any device-internal data anymore. Not sure if that's because of the BMS itself. But the data nodes layout is very similar to that of the MPPT, so that should not be the issue. I will try to find out the reason.

image

Publishing received data via MQTT works very well.

**Comment by Gretel5X** [comment]
> Finally found time to play around with it. Looks really nice!
> 
> With the BMS that is currently attached I can't access any device-internal data anymore. Not sure if that's because of the BMS itself. But the data nodes layout is very similar to that of the MPPT, so that should not be the issue. I will try to find out the reason.
> 
> image
> 
> Publishing received data via MQTT works very well.

This has to do with the way the `info.json` is used and was not introduced with these changes, the same problem is on main already. I would suggest we merge this PR and then fix it

**Pull request merged**